### PR TITLE
Implement visual changes for marketplace plugin cards

### PR DESCRIPTION
**Pull request by michalkleiner**
### Description:

Implement redesign of marketplace plugin cards.

Ref. DEV-17325

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

### CODE REVIEW

**Comment by michalkleiner** [comment]
In principle this is ready for review, there might be an odd UI test needing some more attention. 

Also tried to address a flaky test related to new plugin version update when it shouldn't be checking those in UI tests unless explicitly instructed.

File: plugins/CorePluginsAdmin/Controller.php [hunk 1]
```diff
@@ -222,17 +222,19 @@ private function createPluginsOrThemesView($template, $themesOnly)
...
-                $view->pluginsHavingUpdate    = $this->marketplacePlugins->getPluginsHavingUpdate();
-
-                $view->pluginUpdateNonces = [];
-                foreach ($view->pluginsHavingUpdate as $name => $plugin) {
-                    $view->pluginUpdateNonces[$name] = Nonce::getNonce($plugin['name']);
+                if (!$skipPluginUpdateCheck) {
+                    $view->pluginsHavingUpdate = $this->marketplacePlugins->getPluginsHavingUpdate();
+                    foreach ($view->pluginsHavingUpdate as $name => $plugin) {
+                        $view->pluginUpdateNonces[$name] = Nonce::getNonce($plugin['name']);
+                    }
```
Comment (sgiehl) [inline plugins/CorePluginsAdmin/Controller.php]: Those changes look unneeded as well. `getPluginsHavingUpdate` will already skip any check if the config is set, so there would no need to do that here as well. So either we revert that here, or in `getPluginsHavingUpdate`

File: plugins/Marketplace/Plugins.php [hunk 2]
```diff
@@ -343,4 +358,74 @@ private function addMissingRequirements($plugin)
...
+
+        if (($num >= 1000) && ($num < 100000)) {
+            $nice = round($num / 1000, 1, PHP_ROUND_HALF_DOWN) . 'k';
+        }
+        elseif (($num >= 100000) && ($num < 1000000)) {
+            $nice = floor($num / 100000) . 'k';
+        }
+        elseif ($num >= 1000000) {
+            $nice = floor($num / 1000000) . 'm';
+        }
```
Comment (sgiehl) [inline plugins/Marketplace/Plugins.php:422]: Should we use a shorting of numbers that match the language? We currently don't import such formats from CLDR, but I guess the correct parts would be e.g. https://github.com/unicode-org/cldr-json/blob/main/cldr-json/cldr-numbers-full/main/en/numbers.json#L62-L86

File: config/environment/ui-test.php [hunk 3]
```diff
@@ -16,8 +16,14 @@
...
 
-    // disable check for plugin updates during UI tests
-    'dev.disable_plugin_update_checks' => true,
+    // disable check for plugin updates during UI tests, allow for override
+    'dev.disable_plugin_update_checks' => Piwik\DI::decorate(function ($previous, Container $c) {
+        if ($c->get('test.vars.forceEnablePluginUpdateChecks')) {
+            return false;
+        }
+
+        return true;
```
Comment (sgiehl) [inline config/environment/ui-test.php]: That could be simplified to:
```suggestion
        return !$c->get('test.vars.forceEnablePluginUpdateChecks');
```

File: plugins/Marketplace/Plugins.php [hunk 4]
```diff
@@ -247,7 +254,8 @@ private function enrichPluginInformation($plugin)
         $plugin['isActivated']  = $this->isPluginActivated($plugin['name']);
         $plugin['isInvalid']    = $this->pluginManager->isPluginThirdPartyAndBogus($plugin['name']);
         $plugin['canBeUpdated'] = $plugin['isInstalled'] && $this->hasPluginUpdate($plugin);
-        $plugin['lastUpdated'] = $this->toShortDate($plugin['lastUpdated']);
+        $plugin['lastUpdated']  = $this->toShortDate($plugin['lastUpdated']);
+        $plugin['canBePurchased'] = !$plugin['isDownloadable'] && $plugin['shop'] && ($plugin['shop']['url'] ?? false);
```
Comment (sgiehl) [inline plugins/Marketplace/Plugins.php]: I'd personally prefer
```suggestion
        $plugin['canBePurchased'] = !$plugin['isDownloadable'] && !empty($plugin['shop']['url']);
```

File: plugins/Marketplace/images/previews/Matomo/AbTesting.png
Comment (sgiehl) [inline plugins/Marketplace/images/previews/Matomo/AbTesting.png:1]: I guess we discussed that before: Long term those images need to be placed in the plugins and the marketplace should serve the images. I would personally prefer to directly do that and avoid adding images to the repo that are going to be removed again in the future. This will not only blow up the build size, but also the repo size.

File: plugins/Marketplace/images/matomo-badge.png
Comment (sgiehl) [inline plugins/Marketplace/images/matomo-badge.png:1]: Wouldn't it be more useful to use a SVG instead? might be smaller and could be reused for different sizes later...

File: plugins/Marketplace/tests/UI/Marketplace_spec.js [hunk 5]
```diff
@@ -97,14 +97,15 @@ describe("Marketplace", function () {
...
+        testEnvironment.forceEnablePluginUpdateChecks = forceEnableUpdateChecks ? 1 : 0;
         testEnvironment.save();
     }
 
     ['superuser', 'user', 'multiUserEnvironment'].forEach(function (mode) {
 
         if (mode !== 'user') {
             it('should show available updates in plugins page', async function() {
-                setEnvironment(mode, noLicense);
+                setEnvironment(mode, noLicense, true);
```
Comment (sgiehl) [inline plugins/Marketplace/tests/UI/Marketplace_spec.js]: Just out of curiosity: Why was this change needed, if it didn't cause any test result to change? 🤔

File: plugins/Marketplace/templates/plugin-list.twig [hunk 6]
```diff
@@ -2,151 +2,116 @@
...
+                                <img src="{{ plugin.previewImage }}" alt="" class="preview-image">
+                                <div class="content-container">
+                                    <div class="card-content-top">
+                                        {{ marketplaceMacro.matomoPluginBadge(plugin.owner, 'matomo-badge-top') }}
+                                        <div class="price">
+                                            {% if plugin.canBePurchased and plugin.priceFrom %}
+                                                {{ 'Marketplace_PriceFromPerPeriod'|translate(plugin.priceFrom.prettyPrice, plugin.priceFrom.period) }}
+                                            {% else %}
+                                                {{ 'Marketplace_Free'|translate }}
+                                            {% endif %}
```
Comment (sgiehl) [inline plugins/Marketplace/templates/plugin-list.twig]: I think we should add an additional case here. If you have already added a license key and a premium plugin is included in that license, but isn't installed, the marketplace would now show `Free`. That is not fully correct. Maybe it would be better to show something like `purchased`, `already purchased`, `included in license` or similar.

**Review by sgiehl — CHANGES_REQUESTED** [review]
Some first rough feedback that came to my mind when looking through the code only. Still need to do some testing and check if the new layout is correct in all cases...

File: plugins/Marketplace/tests/UI/Marketplace_spec.js (hunk 5 above)
Comment (michalkleiner) [reply plugins/Marketplace/tests/UI/Marketplace_spec.js]: It works together with the update checks in the backend that were changed in this PR as well, so it still correctly checks for updates here, but otherwise ignore them for all other UI tests.

File: plugins/Marketplace/Plugins.php (hunk 2 above)
Comment (michalkleiner) [reply plugins/Marketplace/Plugins.php:422]: Could do, however how often are the marketplace related texts translated?
We'd need to have a confirmation from design that we can drop the decimal points as the CLDR notation seems to only support integers whereas here we have e.g. `16.3k` as the number. For lower numbers like `1.9k` if we floor'd it to 1k it might be too much of a drop and rounding to 2k is incorrectly adding to our downloads counts.

I'm 50/50 on this so happy to take guidance.

File: plugins/Marketplace/templates/plugin-list.twig (hunk 6 above)
Comment (michalkleiner) [reply plugins/Marketplace/templates/plugin-list.twig]: Yeah, I see what you mean. That would be a functional change as we don't have that covered with the old cards either, but I guess it makes sense. It's "free to install", but may not be covering the fact it's paid and purchased. I'll think about it and talk to Javi what we think is the best wording here.

File: plugins/CorePluginsAdmin/Controller.php (hunk 1 above)
Comment (michalkleiner) [reply plugins/CorePluginsAdmin/Controller.php]: Good spot, will test if this can be removed, but you're likely correct that it can be removed here.

File: plugins/Marketplace/images/matomo-badge.png
Comment (michalkleiner) [reply plugins/Marketplace/images/matomo-badge.png:1]: There might be more work involved in getting a well-structured and scalable SVG — tried one from design team and it was worse than the PNG — so PNG looked easier to use in this space where we don't really need to scale much. But can ask design team for another SVG iteration and use it here if we think it's good value.

File: plugins/Marketplace/images/previews/Matomo/AbTesting.png
Comment (michalkleiner) [reply plugins/Marketplace/images/previews/Matomo/AbTesting.png:1]: Yep, the long term plan is not to have the preview images here (apart from the generic ones). We can have an offline discussion where/how to achieve that within the marketplace itself.

File: plugins/Marketplace/images/matomo-badge.png
Comment (sgiehl) [reply plugins/Marketplace/images/matomo-badge.png:1]: I thought we maybe would already have one. If that's not the case, we don't need to do that as part of this issue.

**Review by sgiehl — CHANGES_REQUESTED** [review]
Here some feedback regarding the layout:

- The license key section completely breaks the layout. Especially when a license key is set, everything below 1250px is broken. Might not have been part of this issue, but without fixing it, the new layout of the cards actually doesn't work well on lower resolutions.
- When a plugin is already installed, the card shows an according information. For some languages and sizes this text might break into multiple lines. It then looks a bit misplaced:
   
image

- There seems to be too much space between the plugin description and the downloads/cta. Seems the plugin name has font-size 24px, while it has 32px on the layouts.

image

**Comment by michalkleiner** [comment]
> * When a plugin is already installed, the card shows an according information. For some languages and sizes this text might break into multiple lines. It then looks a bit misplaced:
> 
> image

Discussed with the designer and this is acceptable, even if it goes onto 3 lines.


> * There seems to be too much space between the plugin description and the downloads/cta. Seems the plugin name has font-size 24px, while it has 32px on the layouts.
> 
> image

This is a change requested by the designer after a review as the dashboard title at the top (Marketplace) was significantly smaller than the card titles.

The space between the title and the description (top section), and the number of downloads and the CTA (bottom section) varies across the list of all plugins since they can have multiline titles, multiline descriptions and the CTA/notification can also span multiple lines in some cases, so the height of the card (already shorter than in the designs) and the space is a balance to accommodate the dynamic content. This has also been reviewed by the designer in various viewport widths and approved.

@Javi-Ormaechea anything to add here from your perspective?

File: plugins/Marketplace/templates/plugin-list.twig (hunk 6 above)
Comment (michalkleiner) [reply plugins/Marketplace/templates/plugin-list.twig]: Discussed with the designer and we will always show the price when the plugin is paid, even when it's installable or already installed. It's an indication of price, not of whether it's been purchased or can be purchased or is for $0.

File: plugins/CorePluginsAdmin/Controller.php (hunk 1 above)
Comment (michalkleiner) [reply plugins/CorePluginsAdmin/Controller.php]: Removed

File: plugins/Marketplace/Plugins.php (hunk 2 above)
Comment (michalkleiner) [reply plugins/Marketplace/Plugins.php:422]: Looked at this again and I'd leave it as is, if that's ok with you.

**Comment by michalkleiner** [comment]
Ready for re-review. The visual and markup tweaks haven't changed for a while, the latest round of updates were related to the CTA button and it correctly showing for different users with different license states (user/admin/multiEnv with/without a licence etc.)

The design looked at this several times and is happy with the responsive behaviour and the card sizes, font sizes and spacing (which slightly differ from the designs after the design met the real world UI).

Plugin images are currently added here but can be refactored in a follow up PR as dealing with that was defined as out of scope for this PR. It can be refactored before next version is released to remove the images from the codebase again (and move to plugins marketplace).

**Review by sgiehl — CHANGES_REQUESTED** [review]
> The license key section completely breaks the layout. Especially when a license key is set, everything below 1250px is broken. Might not have been part of this issue, but without fixing it, the new layout of the cards actually doesn't work well on lower resolutions.

This one is still the case. Adding a `flex-wrap: wrap` to the licenseToolbar element should improve that, as it would allow the buttons to wrap into multiple line.

Regarding the images: My concern here is, we are adding images that are meant to be removed "soonish" again. So we will increase our repository size forever for no "good" reason (unless we rebase and purge out the images from history when we delete them again). Also I'm pretty sure if we don't force to get a clean solution for that now (and serve them through the marketplace), we won't be able to work on that any time this year again.

Since all other concerns have been cleared up by the layout team and it sounds like they have approved the current look, I won't invest further time to check if everything matches the designs exactly.

**Comment by michalkleiner** [comment]
> > The license key section completely breaks the layout. Especially when a license key is set, everything below 1250px is broken. Might not have been part of this issue, but without fixing it, the new layout of the cards actually doesn't work well on lower resolutions.
> 
> This one is still the case. Adding a `flex-wrap: wrap` to the licenseToolbar element should improve that, as it would allow the buttons to wrap into multiple line.

Will recheck this again; I haven't noticed it in the UI tests nor locally, but it's possible I haven't had the license key added when testing smaller breakpoints.

**Comment by michalkleiner** [comment]
> > The license key section completely breaks the layout. Especially when a license key is set, everything below 1250px is broken. Might not have been part of this issue, but without fixing it, the new layout of the cards actually doesn't work well on lower resolutions.
> 
> This one is still the case. Adding a `flex-wrap: wrap` to the licenseToolbar element should improve that, as it would allow the buttons to wrap into multiple line.

Hi @sgiehl, can you please add some screenshots here or to the internal ticket for me to see how broken the layout is for you under 1250px? I can't seem to replicate anything misbehaving in Brave on MacOS. The licence section may be a bit wider in some cases, but that is already the case for me with the existing cards. There is also a plan to remove the licence section from the plugins list in favour of the settings section.

**Comment by michalkleiner** [comment]
> Regarding the images: My concern here is, we are adding images that are meant to be removed "soonish" again. So we will increase our repository size forever for no "good" reason (unless we rebase and purge out the images from history when we delete them again). Also I'm pretty sure if we don't force to get a clean solution for that now (and serve them through the marketplace), we won't be able to work on that any time this year again.

I get the sentiment and I agree that it would be best to address that now rather than later, so I'll ask PO for scope change to the ticket.

What would you see as the best option for hosting the images? The marketplace itself, or a specific file name in the plugins repo, or specific key/entry in plugin.json with a URL of the image? Would that satisfy CSP rules if it's 3rd party URL?

**Comment by sgiehl** [comment]
> Hi @sgiehl, can you please add some screenshots here or to the internal ticket for me to see how broken the layout is for you under 1250px? I can't seem to replicate anything misbehaving in Brave on MacOS. The licence section may be a bit wider in some cases, but that is already the case for me with the existing cards. There is also a plan to remove the licence section from the plugins list in favour of the settings section.

As you can see: The buttons are actually wider as the screen should be and the site becomes scrollable. Add the `flex-wrap: wrap` to the license element, fixes this. This might have already been the case before, but as it's very easy to fix, I guess it would be worth to directly include it.

image

**Comment by sgiehl** [comment]
> What would you see as the best option for hosting the images? The marketplace itself, or a specific file name in the plugins repo, or specific key/entry in plugin.json with a URL of the image? Would that satisfy CSP rules if it's 3rd party URL?

The plugins can already have screenshots. Those screenshots need to be located in a certain directory within the plugin (repo). I think the names of the image might currently be used as the description or similar. We could simply define a fixed name like `previewimage.png`, which would automatically be used as marketplace preview image and excluded from the screenshots or similar. The image would then be hosted on the marketplace, which should already be included in the CSP header.

**Comment by michalkleiner** [comment]
> > Hi @sgiehl, can you please add some screenshots here or to the internal ticket for me to see how broken the layout is for you under 1250px? I can't seem to replicate anything misbehaving in Brave on MacOS. The licence section may be a bit wider in some cases, but that is already the case for me with the existing cards. There is also a plan to remove the licence section from the plugins list in favour of the settings section.
> 
> As you can see: The buttons are actually wider as the screen should be and the site becomes scrollable. Add the `flex-wrap: wrap` to the license element, fixes this. This might have already been the case before, but as it's very easy to fix, I guess it would be worth to directly include it.

Yes, this was behaving like in the image before this change, so I considered it out of scope, but I agree the fix was quite simple and improves the overall usability of the marketplace section, so I've added it. It required tweaking the margins around the input and the buttons to cater for their spacing when they start stacking.

I've also adjusted the alignment of the buttons with the input field and updated relevant UI test screenshots.

This adjustment can be tested/reviewed, @sgiehl.

**Comment by sgiehl** [comment]
@michalkleiner The change works nicely. It now looks a lot better for smaller resolutions 👍

**Review by sgiehl — APPROVED** [review]
This looks fine for me now :+1:
